File: android/app/src/main/kotlin/app/critalarm/alarm/DebugStateRule.kt
package app.critalarm.alarm

/** Pure precedence rule for the diagnostic phone-state label. */
object DebugStateRule {
    data class Input(
        val active: Boolean = false,
        val acknowledged: Boolean = false,
        val closed: Boolean = false,
        val inLocalAckedSet: Boolean = false,
        val live: Boolean = false,
        val rearmPending: Boolean = false,
        val ringUntilMillis: Long? = null,
    )

    fun phoneState(input: Input, nowMillis: Long): String = when {
        input.live -> "ringing"
        input.closed -> "closed"
        input.acknowledged && input.inLocalAckedSet -> "acked here"
        input.acknowledged -> "acked elsewhere"
        input.active && input.rearmPending -> "silenced"
        input.ringUntilMillis?.let { it < nowMillis } == true && !input.rearmPending -> "expired"
        else -> "unknown"
    }
}

File: ios/Shared/Alarm/DebugStateRule.swift
import Foundation

/// Framework-free precedence rule for the diagnostic phone-state label.
enum DebugStateRule {
    struct Input {
        var active = false
        var acknowledged = false
        var closed = false
        var inLocalAckedSet = false
        var live = false
        var rearmPending = false
        var ringUntil: Date?
    }

    static func phoneState(_ input: Input, now: Date) -> String {
        if input.live { return "ringing" }
        if input.closed { return "closed" }
        if input.acknowledged && input.inLocalAckedSet { return "acked here" }
        if input.acknowledged { return "acked elsewhere" }
        if input.active && input.rearmPending { return "silenced" }
        if let ringUntil = input.ringUntil, ringUntil < now, !input.rearmPending { return "expired" }
        return "unknown"
    }
}
